modules/compliance-tests/src/smithy4s/compliancetests/ClientHttpComplianceTestCase.scala

```diff
@@ -0,0 +1,215 @@
+                  toDecode = v,
+                  charset = StandardCharsets.UTF_8,
+                  plusIsSpace = true
+                )
+              )
+          }
+        }.toMap
+      )
+
+    val uriAssert = assert.eql(expectedUri, request.uri)
+    val methodAssert = assert.eql(
+      testCase.method.toLowerCase(),
+      request.method.name.toLowerCase()
+    )
+
+    List(
+      bodyAssert,
+      headerAssert.pure[IO],
+      uriAssert.pure[IO],
+      methodAssert.pure[IO]
+    ).combineAll
+  }
+
+  private[compliancetests] def clientRequestTest[I, E, O, SE, SO](
+      endpoint: Endpoint[Op, I, E, O, SE, SO],
+      testCase: HttpRequestTestCase
+  ): ComplianceTest[IO] = {
+    type R[I_, E_, O_, SE_, SO_] = IO[O_]
+
+    val inputFromDocument = Document.Decoder.fromSchema(endpoint.input)
+    ComplianceTest[IO](
+      name = endpoint.id.toString + "(client|request): " + testCase.id,
+      run = {
+        val input = testCase.params
+          .map { inputFromDocument.decode(_).liftTo[IO] }
+          .getOrElse(IO.pure(().asInstanceOf[I]))
+
+        deferred[Request[IO]].flatMap { requestDeferred =>
+          val app = HttpRoutes
+            .of[IO] { case req =>
+              req.body.compile.toVector
+                .map(fs2.Stream.emits(_))
+                .map(req.withBodyStream(_))
+                .flatMap(requestDeferred.complete(_))
+                .as(Response[IO]())
+            }
+            .orNotFound
+
+          val clientRes = getClient match {
+            case Left(fromApp) => fromApp(app)
+            case Right(fromPort) =>
+              for {
+                port <- retryResource(emberServer(app))
+                client <- fromPort(port)
+              } yield client
+          }
+
+          clientRes.use { client =>
+            // avoid blocking the test forever...
+            val recordedRequest = requestDeferred.get.timeout(1.second)
+
+            input
+              .flatMap { in =>
+                service
+                  .asTransformation[R](client)
+                  .apply(endpoint.wrap(in))
+              }
+              // deal with the empty response generated in the mock
+              .attemptNarrow[PayloadError]
+              .productR(recordedRequest)
+              .flatMap(matchRequest(_, testCase))
+          }
+        }
+      }
+    )
+  }
+
+  def allClientTests(
+  ): List[ComplianceTest[IO]] = {
+    service.endpoints.flatMap { case endpoint =>
+      endpoint.hints
+        .get(HttpRequestTests)
+        .map(_.value)
+        .getOrElse(Nil)
+        .filter(_.protocol == protocolTag.id.toString())
+        .filter(tc => tc.appliesTo.forall(_ == AppliesTo.CLIENT))
+        .map(tc => clientRequestTest(endpoint, tc))
+    }
+  }
+
+  private val randomInt =
+    Resource.eval(IO(scala.util.Random.nextInt(9999)))
+
+  private val randomPort = randomInt.map(_ + 50000)
+
+  private def emberServer(app: HttpApp[IO]): Resource[IO, Int] =
```

Comment:
I wanted to make this abstract lke we did for `PizzaClientSpec`, but this class is a concrete one.

It feels odd to have this kind of interface:

```scala
class ClientHttpComplianceTestCase[...](
...,
    getClient: Either[
      HttpApp[IO] => Resource[IO, smithy4s.Monadic[Alg, IO]],
      Int => Resource[IO, smithy4s.Monadic[Alg, IO]]
    ],
    mkServer: HttpApp[IO] => Resource[IO, Int]
)
```

because we might as well merge these two parameters into 

```scala
    getClient: Either[
      HttpApp[IO] => Resource[IO, smithy4s.Monadic[Alg, IO]],
      (HttpApp[IO] => Resource[IO, Int], Int => Resource[IO, smithy4s.Monadic[Alg, IO]])
    ],
```

which is almost exactly like

```scala
    getClient: Either[
      HttpApp[IO] => Resource[IO, smithy4s.Monadic[Alg, IO]],
      HttpApp[IO] => Resource[IO, Resource[IO, smithy4s.Monadic[Alg, IO]]
    ],
```

which could very well be 

```scala
    getClient: 
      HttpApp[IO] => Resource[IO, smithy4s.Monadic[Alg, IO]],
```

and that doesn't really help us.

How about this: we make this class abstract, make the server method abstract, then provide a concrete implementation that does http4s+ember but we don't use that implementation in our own tests?

In that case I'd actually have a single parameter here:

```scala
    getClient: 
      HttpApp[IO] => Resource[IO, smithy4s.Monadic[Alg, IO]],
```

and make it an `Either` in the http4s-specific implementation.

@Baccata @daddykotex

Reply:
dunno how to make a PR to a fork from a fork, guess I'll just leave it here: https://github.com/kubukoz/smithy4s/tree/compliance-native-fix - this does what I described above.

Reply:
thanks for the branch that helped to understand what you were saying above. I think you should push that to my branch, no worries.

that makes sense to me, and if you can get rid of the duplicated test classes (for compat) go ahead

Reply:
I had to add a `locally` to avoid an unused import turned into a warning

I tried to use `@nowarn` but could not get it to work

Reply:
I'm confused as to the addition of an hard-coupling to Ember, tbh. In particular, doesn't that pose a problem in Native ?

Reply:
It does not (for us) because the ember specific class isn't accessible from the test suite. For others it does (eg for an Akka http client implementation)

Reply:
I think we may want for it to remain abstract. It adds a little bit more burden downstream (like, 10 more lines), but it removes one dependency from the artifact. I do like the 

``` 
getClient: HttpApp[IO] => Resource[IO, smithy4s.Monadic[Alg, IO]]
```

As long as we document it proper, I'm happy to get rid of the Blaze-specific implementation. WDYT ?

Reply:
(jfc why is the resolve button so accessible)

yeah, let's. We can do the http4s thingy in our own tests.

Reply:
> As long as we document it proper

@daddykotex wanna take this one? 😅

Reply:
yeah sure,
let me pick up every detail from this conversation :{

Reply:
as I am writing the docs, I realize: would it make sense for the module to be called `compliance-testket`? maybe even `http-compliance-testkit` or something like that?

Reply:
![localhost_3000_smithy4s_docs_protocols_compliance-tests](https://user-images.githubusercontent.com/5230460/192909731-30f89785-32c2-4d08-a840-fd8d0c9d336f.png)

In the screenshot, the side menu is misaligned, please disregard that, it's the screenshot engine that fails when you trigger the process and you're half way down the page

Reply:
<img width="113" alt="image" src="https://user-images.githubusercontent.com/894884/192915140-e18c850c-3f9d-4ce9-9d6b-4f3f44386ee7.png">

something sus going on here

Reply:
ahh it's because I was on a watch loop, copied the file (so mdoc produced a file) then I renamed it (so mdoc produced a second file, leaving the copy behind in the output dir)

Reply:
fixed 
![Screen Shot 2022-09-29 at 09 49 18](https://user-images.githubusercontent.com/5230460/193049632-2161229a-1466-498d-b05b-ea66a41c7f2a.png)